### PR TITLE
Feature: Add broker for websocket messages

### PR DESCRIPTION
**Pull request by sleiss**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
As we seek for a distributed setup of Artemis (with multiple instances) we have to handle users being connected to different instances but still being able to interact with each other.
This PR adds a broker (based on [Apache ActiveMQ Artemis](https://activemq.apache.org/components/artemis/)) that will forward messages between different instances, especially in the context of WebSockets.

### Description
Artemis will connect to a broker to forward WebSocket messages that could not be delivered to an user directly (because the user is connected to a different node) to the other nodes. Also, messages that don't target a specific user but a group of users will be forwarded.
The other instances will receive these messages and each instance will send the message to the users connected to itself.
Therefore, every user will be reached despite being connected to different nodes.

This PR also adds support for multiple brokers so in case the broker goes down for whatever reason, Artemis will automatically try to connect to the next available broker. After the last broker went down, the first broker will be retried.

If no brokers are configured, the behaviour will not change.

This PR also changes the behaviour of the `ParticipationTeamWebsocketService` so that a distributed map (using Hazelcast) will be used instead of a HashMap. Changes made on one instance will therefore also replicate to other instances.

Also a small bugfix is contained here that fixes a bug where the client of a student would sometimes try to subscribe to an TA/instructor/admin-only topic.

### Steps for Testing
1. Start Artemis locally without making any changes - the behaviour should not change as no second node/broker is present.
2. Now start the ActiveMQ broker using `docker-compose -f src/main/docker/activemq.yml up`. Adjust the config of Artemis so that the value `spring.websocket.broker.addresses` is set to `localhost:61613`.
3. Copy the `Artemis (Server)` IntelliJ run config and add `SPRING_APPLICATION_JSON={"server":{"port":8081}}` to the environment variables input box of the copied configuration. This will start the second instance of Artemis on a different port (8081) to prevent 2 instances on the same port.
image
4. Start both instances of Artemis in IntelliJ.
5. Start two browsers and connect to the 2 different instances (they are reachable under `localhost:8080` and `localhost:8081`) with one browser each (use a different user). Now create a quiz on one instance (instructor) and set it to visible.
6. The quiz should be visible to the user (student) on the second instance (you might have to reload). Click on "Open quiz".
7. Start the quiz with the instructor account -> it should show for the student, despite being connected to a different instance.
NOTE: Submitting might not work, depending on your current setup and whether you are using the registry or not. Submitting quizzes is not part of this PR.

Peek 2020-06-11 23-37

**Testing on the test server**:
The broker is setup on test server 2. You can perform the steps above there, if you use 2 different browsers. nginx is currently configured to connect to the server with the least connections, so if you test alone, you should be connected to both instances once.

### CODE REVIEW

**Comment by sleiss** [comment]
@krusche I added documentation and added the documentation here:
https://confluence.ase.in.tum.de/display/ArTEMiS/ActiveMQ+Message+Broker

File: src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java [hunk 1]
```diff
@@ -118,9 +131,54 @@ public void init() {
 
     @Override
     public void configureMessageBroker(MessageBrokerRegistry config) {
-        config.enableSimpleBroker("/topic").setHeartbeatValue(new long[] { 10000, 20000 }).setTaskScheduler(messageBrokerTaskScheduler);
-        // increase the limit of concurrent connections (default is 1024 which is much too low)
-        // config.setCacheLimit(10000);
+        // Try to create a TCP client that will connect to the message broker (or the message brokers if multiple exists).
+        // If tcpClient, there is no valid address specified in the config. This could be due to a development setup or a mistake in the config.
```
Comment (kloessst) [inline src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java:135]: "If tcpClient is null..."?

Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java]: Good catch!

**Comment by krusche** [comment]
We just found some error logs with respect to the heartbeat (TTL) and to multiple subscriptions. They can probably be ignored, but we want to further investigate them.

@sleiss also just found out that the k6 WebSocket tests won't work any more because the WebSocket message format has changed. We first like to fix the tests and run another simulation with 1500 test users to make sure nothing is broken in terms of performance

**Pull request merged**